plugin.go

```diff
@@ -34,6 +37,149 @@ type Logger interface {
 	NamedLogger(name string) *zap.Logger
 }
 
+type FileServer func(next http.Handler, w http.ResponseWriter, r *http.Request, fp string)
+
+func createMutableServer(s *Plugin) FileServer {
+	return func(next http.Handler, w http.ResponseWriter, r *http.Request, fp string) {
+		// ok, file is not in the forbidden list
+		// Stat it and get file info
+		f, err := s.root.Open(fp)
+		if err != nil {
+			// else no such file, show error in logs only in debug mode
+			s.log.Debug("no such file or directory", zap.Error(err))
+			// pass request to the worker
+			next.ServeHTTP(w, r)
+			return
+		}
+
+		// at high confidence here should not be an error
+		// because we stat-ed the path previously and know, that that is file (not a dir), and it exists
+		finfo, err := f.Stat()
+		if err != nil {
+			// else no such file, show error in logs only in debug mode
+			s.log.Debug("no such file or directory", zap.Error(err))
+			// pass request to the worker
+			next.ServeHTTP(w, r)
+			return
+		}
+
+		defer func() {
+			err = f.Close()
+			if err != nil {
+				s.log.Error("file close error", zap.Error(err))
+			}
+		}()
+
+		// if provided path to the dir, do not serve the dir, but pass the request to the worker
+		if finfo.IsDir() {
+			s.log.Debug("possible path to dir provided")
+			// pass request to the worker
+			next.ServeHTTP(w, r)
+			return
+		}
+
+		// set etag
+		if s.cfg.CalculateEtag {
+			SetEtag(w, calculateEtag(s.cfg.Weak, f, finfo.Name()))
+		}
+
+		if s.cfg.Request != nil {
+			for k, v := range s.cfg.Request {
+				r.Header.Add(k, v)
+			}
+		}
+
+		if s.cfg.Response != nil {
+			for k, v := range s.cfg.Response {
+				w.Header().Set(k, v)
+			}
+		}
+
+		// we passed all checks - serve the file
+		http.ServeContent(w, r, finfo.Name(), finfo.ModTime(), f)
+	}
+}
+
+type ScannedFile struct {
+	file    http.File
+	name    string
+	modTime time.Time
+	etag    string
+}
+
+func createImmutableServer(s *Plugin) (FileServer, error) {
```

Comment:
Passing plugin pointer is a bad idea. Consider approach without that.

Reply:
Like that?